operator/internal/certrotation/cabundle.go

```diff
@@ -0,0 +1,89 @@
+package certrotation
+
+import (
+	"crypto/x509"
+	"reflect"
+
+	"github.com/openshift/library-go/pkg/crypto"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/util/cert"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// BuildCABundle returns a ConfigMap including all known non-expired signing CAs across rotations.
+func BuildCABundle(opts *Options) (client.Object, error) {
+	cm := newConfigMap(*opts)
+
+	certs, err := manageCABundleConfigMap(cm, opts.Signer.RawCA.Config.Certs[0])
+	if err != nil {
+		return nil, err
+	}
+
+	opts.RawCACerts = certs
+
+	return cm, nil
+}
+
+func newConfigMap(opts Options) *corev1.ConfigMap {
+	current := opts.CABundle.DeepCopy()
+
+	cm := &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      CABundleName(opts.StackName),
+			Namespace: opts.StackNamespace,
+		},
+	}
+
+	if current != nil {
+		cm.Annotations = current.Annotations
+		cm.Labels = current.Labels
+		cm.Data = current.Data
+	}
+
+	return cm
+}
+
+// manageCABundleConfigMap adds the new certificate to the list of cabundles, eliminates duplicates, and prunes the list of expired
+// certs to trust as signers
+func manageCABundleConfigMap(caBundleConfigMap *corev1.ConfigMap, currentSigner *x509.Certificate) ([]*x509.Certificate, error) {
+	if caBundleConfigMap.Data == nil {
+		caBundleConfigMap.Data = map[string]string{}
+	}
+
+	certificates := []*x509.Certificate{}
+	caBundle := caBundleConfigMap.Data[CAFile]
+	if len(caBundle) > 0 {
+		var err error
+		certificates, err = cert.ParseCertsPEM([]byte(caBundle))
+		if err != nil {
+			return nil, err
+		}
+	}
+	certificates = append([]*x509.Certificate{currentSigner}, certificates...)
+	certificates = crypto.FilterExpiredCerts(certificates...)
+
+	finalCertificates := []*x509.Certificate{}
+	// now check for duplicates. n^2, but super simple
+	for i := range certificates {
+		found := false
+		for j := range finalCertificates {
+			if reflect.DeepEqual(certificates[i].Raw, finalCertificates[j].Raw) {
+				found = true
+				break
+			}
+		}
+		if !found {
+			finalCertificates = append(finalCertificates, certificates[i])
+		}
+	}
+
+	caBytes, err := crypto.EncodeCertificates(finalCertificates...)
+	if err != nil {
+		return nil, err
+	}
+
+	caBundleConfigMap.Data[CAFile] = string(caBytes)
```

Comment:
I think I would move the side-effect of modifying the `caBundleConfigMap` out of this function into the enclosing one and use this function just for the parsing / filtering.